extensions/visualizations/oppia-visualization-click-hexbins.directive.ts

```diff
@@ -0,0 +1,158 @@
+          const ctrl = this;
+          const data = HtmlEscaperService.escapedJsonToObj(
+            $attrs.escapedData);
+          const options = HtmlEscaperService.escapedJsonToObj(
+            $attrs.escapedOptions);
+          const imageAndRegions = HtmlEscaperService.escapedJsonToObj(
+            $attrs.imageAndRegionsWithValue);
+
+          const showTooltip = (hexBin: IHexBin) => {
+            if (hexBin.length === 0) {
+              return;
+            }
+            d3.select('#click-hexbin-chart-tooltip')
+              .style('visibility', 'visible')
+              .style('left', `${hexBin.x}px`)
+              .style('top', `${hexBin.y + 16}px`)
+              .text(`${hexBin.length} click${hexBin.length > 1 ? 's' : ''}`);
+          };
+
+          const moveTooltip = (hexBin: IHexBin) => {
+            if (hexBin.length === 0) {
+              return;
+            }
+            d3.select('#click-hexbin-chart-tooltip')
+              .style('visibility', 'visible');
+          };
+
+          const hideTooltip = (hexBin: IHexBin) => {
+            d3.select('#click-hexbin-chart-tooltip')
+              .style('visibility', 'hidden');
+          };
+
+          const imageUrl = AssetsBackendApiService.getImageUrlForPreview(
+            ContextService.getEntityType(), ContextService.getEntityId(),
+            imageAndRegions.imagePath);
+          const imageDimensions = ImagePreloaderService.getDimensionsOfImage(
+            imageAndRegions.imagePath);
+          const imageAspectRatio = (
+            imageDimensions.height / imageDimensions.width);
+
+          const containerWidth = $('#click-hexbin-chart').width();
+          const containerHeight = Math.round(containerWidth * imageAspectRatio);
+
+          ctrl.$onInit = function() {
+            d3.select('#click-hexbin-chart-tooltip')
+              .style('visibility', 'hidden');
+
+            // Define the data for the hexbins.
+            const hexbin = d3Hexbin.hexbin()
+              .x(d => (d.answer.clickPosition[0] * containerWidth))
+              .y(d => (d.answer.clickPosition[1] * containerHeight))
+              .radius(16)
+              .size([containerWidth, containerHeight]);
+            const bins = hexbin(data);
+
+            // Define the color of hexbins, using the number of grouped clicks
+            // to change their opacity.
+            const rgbaScaleEndPoints = [
+              'rgba(255,255,255,0.25)',
+              'rgba(255,255,255,0.75)'
+            ];
+            const color = d3.scaleLinear()
+              .domain([0, d3.max(bins, (b: IHexBin) => b.length)])
+              // NOTE TO DEVELOPERS: the range type is wrong; rgba string-values
+              // are supported.
+              // @ts-ignore
+              .range(rgbaScaleEndPoints);
+
+            // Construct and add the SVG element for holding the hexbin graph.
+            const svg = d3.select('#click-hexbin-chart').append('svg')
```

Comment:
Instead of manipulating elements in the JS, can you do this declaratively using Angular? I believe we already do similar things e.g. in extensions/interactions/GraphInput.

Reply:
I don't think so, seems to be a limitation of D3 (which GraphInput doesn't seem to use).